### PR TITLE
#3897: Unified connection integration - sql connection improvements

### PR DESCRIPTION
**Pull request by rajmusuku**
- 'Attach to' dropdown would be defaulted to 'Select connection' if sql(only) connection is selected for Spark kernels
    - User will have the ability to add a new connection
    - User can toggle existing connections
- Add 'Select connection' to dropdown if an invalid connection is selected. In this case sql connection.
- Remove 'Select connection' from dropdown if a valid bigdata connection is selected.
- Show error message to user upon selecting a sql connection for Spark kernels

### CODE REVIEW

File: src/sql/parts/notebook/notebookActions.ts [hunk 1]
```diff
@@ -229,33 +232,74 @@ export class AttachToDropdown extends SelectBox {
...
-				let currentKernelSpec = this.model.specs.kernels.find(kernel => kernel.name && kernel.name.toLowerCase() === nameLower);
-				this.loadAttachToDropdown(this.model, currentKernelSpec.display_name);
+			let kernelDisplayName: string = this.getKernelDisplayName();
+			if (kernelDisplayName) {
+				this.loadAttachToDropdown(this.model, kernelDisplayName);
 			}
 		});
 	}
 
+	private updateAttachtoDropdown(model: INotebookModel): void {
```
Comment (corivera) [inline src/sql/parts/notebook/notebookActions.ts]: Need to capitalize the "to" in the middle.

Comment (rajmusuku) [reply src/sql/parts/notebook/notebookActions.ts]: I was considering 'Attachto' as a single word from a usability standpoint in Notebooks.
@kevcunnane @chlafreniere - What do you think?

Comment (kevcunnane) [reply src/sql/parts/notebook/notebookActions.ts]: Yeah it's 2 words. Just like in the UI it's `Attach To:` so it should be camel-cased everywhere in the code

File: src/sql/parts/notebook/notebookActions.ts [hunk 2]
```diff
@@ -229,33 +232,74 @@ export class AttachToDropdown extends SelectBox {
...
+		this.model = model;
+		model.onValidConnectionSelected(validConnection => {
+			let kernelDisplayName: string = this.getKernelDisplayName();
+			if (kernelDisplayName) {
+				if (!validConnection) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/notebookActions.ts]: The code is the exact same, can'y you just invert the validConnection value?
```suggestion
				this.loadAttachToDropdown(this.model, kernelDisplayName, !validConnection);
```

File: src/sql/parts/notebook/notebookActions.ts [hunk 3]
```diff
@@ -229,33 +232,74 @@ export class AttachToDropdown extends SelectBox {
...
+					this.loadAttachToDropdown(this.model, kernelDisplayName, false);
+				}
+			}
+		});
+	}
+
+	private getKernelDisplayName(): string {
+		let kernelDisplayName: string;
+		if (this.model.clientSession.kernel && this.model.clientSession.kernel.name) {
+			let currentKernelName = this.model.clientSession.kernel.name.toLowerCase();
```
Comment (kevcunnane) [inline src/sql/parts/notebook/notebookActions.ts]: Do we need to guard against `this.model.clientSession` and `this.model.clientSession.kernel` being undefined?

File: src/sql/parts/notebook/notebookActions.ts [hunk 4]
```diff
@@ -229,33 +232,74 @@ export class AttachToDropdown extends SelectBox {
...
+			if (currentKernelSpec) {
+				kernelDisplayName = currentKernelSpec.display_name;
+			}
+		}
+		return kernelDisplayName;
+	}
+
 	// Load "Attach To" dropdown with the values corresponding to Kernel dropdown
-	public async loadAttachToDropdown(model: INotebookModel, currentKernel: string): Promise<void> {
+	public async loadAttachToDropdown(model: INotebookModel, currentKernel: string, defaultSelectConnection?: boolean): Promise<void> {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/notebookActions.ts]: Consider a better name here such as
```suggestion
	public async loadAttachToDropdown(model: INotebookModel, currentKernel: string, showSelectConnectionOption?: boolean): Promise<void> {
```

File: src/sql/parts/notebook/notebookActions.ts (hunk 1 above)
Comment (rajmusuku) [reply src/sql/parts/notebook/notebookActions.ts]: Spent about 2 minutes to think while coding it yesterday. Should have gone straight :)

File: src/sql/parts/notebook/notebookActions.ts (hunk 3 above)
Comment (rajmusuku) [reply src/sql/parts/notebook/notebookActions.ts]: Make sense. Was using the exisitng ones..

**Pull request merged**